File: pset01/forge.go
package main

import (
	"encoding/hex"
	"fmt"
	"runtime"
	"time"
)

/*
A note about the provided keys and signatures:
the provided pubkey and signature, as well as "HexTo___" functions may not work
with all the different implementations people could built.  Specifically, they
are tied to an endian-ness.  If, for example, you decided to encode your public
keys as (according to the diagram in the slides) up to down, then left to right:
<bit 0, row 0> <bit 0, row 1> <bit 1, row 0> <bit 1, row 1> ...

then it won't work with the public key provided here, because it was encoded as
<bit 0, row 0> <bit 1, row 0> <bit 2, row 0> ... <bit 255, row 0> <bit 0, row 1> ...
(left to right, then up to down)

so while in class I said that any decisions like this would work as long as they
were consistent... that's not actually the case!  Because your functions will
need to use the same ordering as the ones I wrote in order to create the signatures
here.  I used what I thought was the most straightforward / simplest encoding, but
endian-ness is something of a tabs-vs-spaces thing that people like to argue
about :).

So for clarity, and since it's not that obvious from the HexTo___ decoding
functions, here's the order used:

secret keys and public keys:
all 256 elements of row 0, most significant bit to least significant bit
(big endian) followed by all 256 elements of row 1.  Total of 512 blocks
of 32 bytes each, for 16384 bytes.
For an efficient check of a bit within a [32]byte array using this ordering,
you can use:
    arr[i/8]>>(7-(i%8)))&0x01
where arr[] is the byte array, and i is the bit number; i=0 is left-most, and
i=255 is right-most.  The above statement will return a 1 or a 0 depending on
what's at that bit location.

Messages: messages are encoded the same way the sha256 function outputs, so
nothing to choose there.

Signatures: Signatures are also read left to right, MSB to LSB, with 256 blocks
of 32 bytes each, for a total of 8192 bytes.  There is no indication of whether
the provided preimage is from the 0-row or the 1-row; the accompanying message
hash can be used instead, or both can be tried.  This again interprets the message
hash in big-endian format, where
    message[i/8]>>(7-(i%8)))&0x01
can be used to determine which preimage block to reveal, where message[] is the
message to be signed, and i is the sequence of bits in the message, and blocks
in the signature.

Hopefully people don't have trouble with different encoding schemes.  If you
really want to use your own method which you find easier to work with or more
intuitive, that's OK!  You will need to re-encode the key and signatures provided
in signatures.go to match your ordering so that they are valid signatures with
your system.  This is probably more work though and I recommend using the big
endian encoding described here.

*/

// Forge is the forgery function, to be filled in and completed.  This is a trickier
// part of the assignment which will require the computer to do a bit of work.
// It's possible for a single core or single thread to complete this in a reasonable
// amount of time, but may be worthwhile to write multithreaded code to take
// advantage of multi-core CPUs.  For programmers familiar with multithreaded code
// in golang, the time spent on parallelizing this code will be more than offset by
// the CPU time speedup.  For programmers with access to 2-core or below CPUs, or
// who are less familiar with multithreaded code, the time taken in programming may
// exceed the CPU time saved.  Still, it's all about learning.
// The Forge() function doesn't take any inputs; the inputs are all hard-coded into
// the function which is a little ugly but works OK in this assigment.
// The input public key and signatures are provided in the "signatures.go" file and
// the code to convert those into the appropriate data structures is filled in
// already.
// Your job is to have this function return two things: A string containing the
// substring "forge" as well as your name or email-address, and a valid signature
// on the hash of that ascii string message, from the pubkey provided in the
// signatures.go file.
// The Forge function is tested by TestForgery() in forge_test.go, so if you
// run "go test" and everything passes, you should be all set.
func Forge() (string, Signature, error) {
	// decode pubkey, all 4 signatures into usable structures from hex strings
	pub, err := HexToPubkey(hexPubkey1)
	if err != nil {
		panic(err)
	}

	sig1, err := HexToSignature(hexSignature1)
	if err != nil {
		panic(err)
	}
	sig2, err := HexToSignature(hexSignature2)
	if err != nil {
		panic(err)
	}
	sig3, err := HexToSignature(hexSignature3)
	if err != nil {
		panic(err)
	}
	sig4, err := HexToSignature(hexSignature4)
	if err != nil {
		panic(err)
	}

	var sigslice []Signature
	sigslice = append(sigslice, sig1)
	sigslice = append(sigslice, sig2)
	sigslice = append(sigslice, sig3)
	sigslice = append(sigslice, sig4)

	var msgslice []Message

	msgslice = append(msgslice, GetMessageFromString("1"))
	msgslice = append(msgslice, GetMessageFromString("2"))
	msgslice = append(msgslice, GetMessageFromString("3"))
	msgslice = append(msgslice, GetMessageFromString("4"))

	fmt.Printf("ok 1: %v\n", Verify(msgslice[0], pub, sig1))
	fmt.Printf("ok 2: %v\n", Verify(msgslice[1], pub, sig2))
	fmt.Printf("ok 3: %v\n", Verify(msgslice[2], pub, sig3))
	fmt.Printf("ok 4: %v\n", Verify(msgslice[3], pub, sig4))

	msgStringPre := "forge houstonj2013 2024-02-16 "

	startTime := time.Now()

	var useNewMsg string
	fmt.Printf("The current message is %s", msgStringPre)
	fmt.Println("Do you want to use new message: (Yes/No):")
	fmt.Scanln(&useNewMsg)
	if useNewMsg == "Yes" || useNewMsg == "yes" {
		fmt.Println("Please enter the new message: ")
		fmt.Scanln(&msgStringPre)
		fmt.Println("The new message is :" + msgStringPre)
	} else {
		fmt.Println("Still use current message:" + msgStringPre)
	}

	// your code here!
	// ==
	// Get the secret keys from known signature
	// Randomly guess the unknown secret keys and verify the message and public keys
	// Add nounces into my message as a variance in the random guess

	// get the known screate keys and save in a map; map key from 0 - 511,

	known_bits := make([][]bool, 256)
	sig_index := make([]int, 512)
	for i := 0; i < 512; i++ {
		sig_index[i] = -1
	}
	for i := 0; i < 256; i++ {
		known_bits[i] = make([]bool, 2)
		known_bits[i][0] = false
		known_bits[i][1] = false
	}
	for mi, Imsg := range msgslice {
		for i := 0; i < 256; i++ {
			if Imsg[i/8]>>(7-(i%8))&0x01 == 1 {
				known_bits[i][1] = true
				sig_index[i+256] = mi
			} else {
				known_bits[i][0] = true
				sig_index[i] = mi
			}
		}
	}
	var zeroConstraints []int
	var oneConstraints []int
	for i := 0; i < 256; i++ {
		if known_bits[i][0] == false && known_bits[i][1] == true {
			oneConstraints = append(oneConstraints, i)
		} else if known_bits[i][1] == false && known_bits[i][0] == true {
			zeroConstraints = append(zeroConstraints, i)
		}
	}
	fmt.Printf("zero constaints %d and one constraints %d \n", len(zeroConstraints), len(oneConstraints))
	fmt.Println(zeroConstraints)
	fmt.Println(oneConstraints)

	// Explore the unknown secret keys and messages to forge a signature
	foundMessageString := "forge houstonj2013 2024-02-16 w 357 + 4278692"
	var sig Signature
	var useSavedForge string
	fmt.Printf("The current saved forge is %s \n", foundMessageString)
	fmt.Println("Do you want to use the saved forge: (Yes/No):")
	fmt.Scanln(&useSavedForge)
	if useSavedForge == "Yes" || useSavedForge == "yes" {
		foundmsg := GetMessageFromString(foundMessageString)
		for i := 0; i < 256; i++ {
			if foundmsg[i/8]>>(7-(i%8))&0x01 == 1 && sig_index[i+256] != -1 {
				sig.Preimage[i] = sigslice[sig_index[i+256]].Preimage[i]
			} else if foundmsg[i/8]>>(7-(i%8))&0x01 == 0 && sig_index[i] != -1 {
				sig.Preimage[i] = sigslice[sig_index[i]].Preimage[i]
			} else {
				break
			}
		}
		duration := time.Since(startTime)
		fmt.Println(duration)
		if Verify(foundmsg, pub, sig) {
			fmt.Printf("%s is verified  %v\n", hex.EncodeToString(foundmsg[:]), true)
			return foundMessageString, sig, nil
		} else {
			return "wrong forge", sig, nil
		}
	} else {
		var newString string
		const numJobs = 1000
		sigString := make(chan string, numJobs)
		fmt.Printf("lanching %v rountines to forge the signature \n", numJobs)
		fmt.Println("Version", runtime.Version())
		fmt.Println("NumCPU", runtime.NumCPU())
		fmt.Println("GOMAXPROCS", runtime.GOMAXPROCS(0))
		for w := 1; w <= numJobs; w++ {
			go forgeworker(w, zeroConstraints, oneConstraints, msgStringPre, sigString)
		}
		// hold until the any gorountine return the channel values
		newString = <-sigString
		foundmsg := GetMessageFromString(newString)
		for i := 0; i < 256; i++ {
			if foundmsg[i/8]>>(7-(i%8))&0x01 == 1 && sig_index[i+256] != -1 {
				sig.Preimage[i] = sigslice[sig_index[i+256]].Preimage[i]
			} else if foundmsg[i/8]>>(7-(i%8))&0x01 == 0 && sig_index[i] != -1 {
				sig.Preimage[i] = sigslice[sig_index[i]].Preimage[i]
			} else {
				break
			}
		}
		duration := time.Since(startTime)
		fmt.Println(duration)
		if Verify(foundmsg, pub, sig) {
			fmt.Printf("%s is verified  %v\n", newString, true)
			return newString, sig, nil
		} else {
			return "wrong forge", sig, nil
		}
	}

}

func forgeworker(
	workderId int,
	zeroConstraints []int,
	oneConstraints []int,
	msgStringPre string,
	sigString chan<- string) {
	nounce_pre := fmt.Sprintf("w %d + ", workderId)
	fmt.Printf("Start worker %v \n", workderId)
	var numTries uint64 = 0
	for {
		numTries++
		if numTries%5000000 == 0 {
			fmt.Printf("Worker %v has tried %v times \n", workderId, numTries)
		}

		nouncesS := nounce_pre + fmt.Sprintf("%d", numTries)
		msgString := msgStringPre
		msgString = msgStringPre + nouncesS
		// fmt.Printf("Current string %s \n", msgString)
		mymsg := GetMessageFromString(msgString)

		foundforge := true
		for _, zeroIndex := range zeroConstraints {
			if mymsg[zeroIndex/8]>>(7-(zeroIndex%8))&0x01 == 1 {
				foundforge = false
				break
			}
		}
		if !foundforge {
			continue
		}
		for _, oneIndex := range oneConstraints {
			if mymsg[oneIndex/8]>>(7-(oneIndex%8))&0x01 == 0 {
				foundforge = false
				break
			}
		}
		if foundforge {
			// return the verified string and signature
			fmt.Printf("Find a verified signature and message %s \n", msgString)
			sigString <- msgString
			break
		}

	}
}

// hint:
// arr[i/8]>>(7-(i%8)))&0x01

File: pset01_cuda/forge.cpp
#include <iostream>
#include <iomanip>
#include <cstring>
#include <vector>
#include <atomic>
#include <cstdio>
#include <thread>
#include <unordered_map>
#include <chrono>
#include "lamport.h"
#include "signatures.h"
using namespace std::chrono;

using std::cout;
using std::endl;
using std::string;

Signature forgeSig(Message forge_msg,
                   int sigIndex[],
                   std::vector<Signature> &sigs,
                   std::vector<Message> &msgs)
{
    Signature sig;
    for (int i = 0; i < 256; i++)
    {
        if ((forge_msg[i / 8] >> (7 - (i % 8)) & 0x01) == 1 && sigIndex[i + 256] != -1)
        {
            sig.Preimage[i] = sigs[sigIndex[i + 256]].Preimage[i];
        }
        else if ((forge_msg[i / 8] >> (7 - (i % 8)) & 0x01) == 0 && sigIndex[i + 256] != -1)
        {
            sig.Preimage[i] = sigs[sigIndex[i]].Preimage[i];
        }
        else
        {
            std::cout << "Won't be able to forge signature with the given message" << std::endl;
            break;
        }
    }
    return sig;
}

bool verifyforgeSpeed(Message mymsg,
                      std::vector<int> &zeroConstraints,
                      std::vector<int> &oneConstraints)
{
    for (auto zeroIndex : zeroConstraints)
    {
        if ((mymsg[zeroIndex / 8] >> (7 - (zeroIndex % 8)) & 0x01) == 1)
        {
            // std::cout << "zero " << zeroIndex << " failed " << std::endl;
            return false;
        }
    }

    for (auto oneIndex : oneConstraints)
    {
        if ((mymsg[oneIndex / 8] >> (7 - (oneIndex % 8)) & 0x01) == 0)
        {
            // std::cout << "one " << oneIndex << " failed " << std::endl;
            return false;
        }
    }

    return true;
}

void ForgeWorker(
    int worker_id,
    const std::string forge_string_pre,
    std::vector<int> &zeroConstraints,
    std::vector<int> &oneConstraints,
    std::string &forge_string,
    std::atomic<bool> *found_forge)
{
    unsigned long numTries = 0;
    string nounce_pre = "w " + std::to_string(worker_id) + " + ";
    string trial;
    bool foundForge = false;
    Message mymsg;
    std::cout << "start worker " << worker_id << std::endl;

    while (true)
    {

        numTries++;
        if (numTries % 5000000 == 0)
        {
            std::cout << "workder " << worker_id << " reached " << numTries << " trials" << std::endl;
        }

        trial = forge_string_pre + nounce_pre + std::to_string(numTries);
        mymsg = GetMessageFromString(trial);
        foundForge = verifyforgeSpeed(mymsg, zeroConstraints, oneConstraints);
        if (foundForge)
        {
            (*found_forge).store(true);
            forge_string = trial;
            break;
        }

        if (numTries % 100 == 0)
        {
            if ((*found_forge).load())
            {
                break;
            }
        }
    }
}

int main(int argc, char *argv[])
{

    // start time
    auto start = high_resolution_clock::now();

    // Load and check given public key and signatures
    std::unordered_map<std::string, std::string> global_variables = globalVariables();
    PublicKey pub = HexToPublickey(global_variables["hexPubkey1"]);
    Signature sig1 = HexToSignature(global_variables["hexSignature1"]);
    Signature sig2 = HexToSignature(global_variables["hexSignature2"]);
    Signature sig3 = HexToSignature(global_variables["hexSignature3"]);
    Signature sig4 = HexToSignature(global_variables["hexSignature4"]);
    std::vector<Signature> sigs;
    sigs.push_back(sig1);
    sigs.push_back(sig2);
    sigs.push_back(sig3);
    sigs.push_back(sig4);

    Message msg1 = GetMessageFromString(global_variables["msg1_string"]);
    Message msg2 = GetMessageFromString(global_variables["msg2_string"]);
    Message msg3 = GetMessageFromString(global_variables["msg3_string"]);
    Message msg4 = GetMessageFromString(global_variables["msg4_string"]);
    std::vector<Message> msgs;
    msgs.push_back(msg1);
    msgs.push_back(msg2);
    msgs.push_back(msg3);
    msgs.push_back(msg4);

    if (Verify(msg1, pub, sig1))
    {
        std::cout << "sig 1 is verified" << std::endl;
    }
    else
    {
        std::cout << "sig 1 failed to verify" << std::endl;
    }
    if (Verify(msg2, pub, sig2))
    {
        std::cout << "sig 2 is verified" << std::endl;
    }
    else
    {
        std::cout << "sig 2 failed to verify" << std::endl;
    }
    if (Verify(msg3, pub, sig3))
    {
        std::cout << "sig 3 is verified" << std::endl;
    }
    else
    {
        std::cout << "sig 3 failed to verify" << std::endl;
    }
    if (Verify(msg4, pub, sig4))
    {
        std::cout << "sig 4 is verified" << std::endl;
    }
    else
    {
        std::cout << "sig 4 failed to verify" << std::endl;
    }

    // invidividualized message
    std::string msgStringPre = "forge houstonj2013 2024-02-16 ";
    std::cout << "The current message is " << msgStringPre << std::endl;
    std::cout << "Do you want to use new message : (Yes/No)" << std::endl;
    string useNewMsg;
    std::cin >> useNewMsg;
    if (useNewMsg == "Yes" || useNewMsg == "yes")
    {
        std::cout << "Please enter the new message: " << std::endl;
        std::cin >> msgStringPre;
        std::cout << "The new message is : " + msgStringPre << std::endl;
    }
    else
    {
        std::cout << "Still use current message: " << msgStringPre << std::endl;
    }

    std::vector<std::pair<bool, bool>> known_bits(256, {false, false});
    int sigIndex[512] = {-1};
    for (int si = 0; si < msgs.size(); si++)
    {
        Message tempMsg = msgs[si];
        for (int i = 0; i < 256; i++)
        {
            if ((tempMsg[i / 8] >> (7 - (i % 8)) & 0x01) == 1)
            {
                known_bits[i].second = true;
                sigIndex[i + 256] = si;
            }
            else
            {
                known_bits[i].first = true;
                sigIndex[i] = si;
            }
        }
    }
    std::vector<int> zeroConstraints, oneConstraints;
    for (int i = 0; i < 256; i++)
    {
        if (!known_bits[i].first && known_bits[i].second)
            oneConstraints.push_back(i);
        if (known_bits[i].first && !known_bits[i].second)
            zeroConstraints.push_back(i);
    }

    std::cout << "There are " << oneConstraints.size() << " one constraints and " << zeroConstraints.size() << " zero constraints " << std::endl;
    for (auto z : zeroConstraints)
        std::cout << z << " ";
    std::cout << std::endl;
    for (auto z : oneConstraints)
        std::cout << z << " ";
    std::cout << std::endl;

    std::string foundMessageString = "forge houstonj2013 2024-02-16 w 357 + 4278692";

    std::cout << "The saved forge message is " << foundMessageString << std::endl;
    std::cout << "Do you want to use the saved forge: (Yes/No):" << std::endl;
    string useSavedForge;
    std::cin >> useSavedForge;

    if (useSavedForge == "Yes" || useSavedForge == "yes")
    {
        Message foundMsg = GetMessageFromString(foundMessageString);
        std::cout << "The found message is: " << foundMsg.ToHex() << std::endl;
        if (verifyforgeSpeed(foundMsg, zeroConstraints, oneConstraints))
        {
            std::cout << "The found message is verified by speed check" << std::endl;
        }
        Signature sig = forgeSig(foundMsg, sigIndex, sigs, msgs);

        if (Verify(foundMsg, pub, sig))
        {
            std::cout << foundMessageString << " was found to be able to forge signature" << std::endl;
        }
        else
        {
            std::cout << "saved forge " + foundMessageString << " can't be verified" << std::endl;
        }
    }
    else
    {
        // Speed up search
        const int num_threads = 1000;
        std::vector<std::thread> threads;
        std::string forge_string;
        std::atomic<bool> found_forge{false};

        for (int i = 0; i < num_threads; i++)
        {
            threads.push_back(std::thread(&ForgeWorker,
                                          i,
                                          msgStringPre,
                                          std::ref(zeroConstraints),
                                          std::ref(oneConstraints),
                                          std::ref(forge_string),
                                          &found_forge));
        }
        for (auto &th : threads)
        {
            th.join();
        }
        if (found_forge.load())
        {
            Message foundMsg = GetMessageFromString(forge_string);
            Signature sig = forgeSig(foundMsg, sigIndex, sigs, msgs);

            if (Verify(foundMsg, pub, sig))
            {
                std::cout << "forge signature string found: " << forge_string << std::endl;
            }
        }
        auto stop = high_resolution_clock::now();
        auto duration = duration_cast<seconds>(stop - start);
        cout << "program takes " << duration.count() << " seconds to run" << endl;
    }

    return 0;
}
